### ISSUE TITLE
`atan(tan(pi/2))` = `-pi/2` using `f32`

### ISSUE THREAD

**Issue by CanadaHonk** [rust-lang/rust#108769]
Doing the math expression `atan(tan(pi/2))` using `f32` gives `-pi/2` instead of `pi/2`.

I tried this code:

```rust
fn main() {
    println!("f32: {}", (std::f32::consts::PI / 2.0).tan().atan());
    println!("f64: {}", (std::f64::consts::PI / 2.0f64).tan().atan());
}
```

I expected to see this happen:
```
f32: 1.5707963
f64: 1.5707963267948966
```

Instead, this happened:
```
f32: -1.5707963
f64: 1.5707963267948966
```

### Meta
Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7472c517d73a8e6ab0702feb2f8149d9

**Comment by emilio** [comment]
This also happens in C/C++ (haven't debugged why yet):

```c++
#include <cmath>
#include <cstdio>

int main() {
  printf("%f\n", std::atan(std::tan(float(M_PI / 2))));
  printf("%f\n", std::atan(std::tan(double(M_PI / 2))));
}
```

**Comment by yellowsink** [comment]
I am not sure what specs apply, but given that PI/2 is an undefined value of tan, I'm not sure what the best move actually is here
![image](https://user-images.githubusercontent.com/19270622/222957179-336c939b-2de6-4411-b576-b131e24c2168.png)

**Comment by workingjubilee** [comment]
Clearly, we reimplement `f32::tan` and have it return NaN on this input, because programmers like correctly representing undefined values, right? :smile_cat:

**Comment by 823984418** [comment]
> right? 😸

You must be joking.

IEEE754 doesn't think so?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bd7bd05e3069a3fbb290801df197ac1d

**Comment by yellowsink** [comment]
> > right? smile_cat
> 
> You must be joking.
> 
> IEEE754 doesn't think so?
> 
> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bd7bd05e3069a3fbb290801df197ac1d

Floats should always be compared with a tolerance value for a reason

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3976672b25d566efd5f8cf5d24a030a7

**Comment by 823984418** [comment]
> Floats should always be compared with a tolerance value for a reason

Of course, but it is clear that the questioner does not necessarily agree with it...

**Comment by yellowsink** [comment]
This seems fairly standard behaviour across a few languages I've tested, here's dotnet for example:

```fs
> tan (System.Math.PI / 2.);;
val it: float = 1.633123935e+16

> tan (float32 (System.Math.PI / 2.));;
val it: float32 = -22877332.0f
```

**Comment by 823984418** [comment]
Yes, IEEE 754 indirectly stipulates it.

**Comment by yellowsink** [comment]
Well given that, this would seem a case of "floating point is a pain" not "this is a bug"?

**Comment by workingjubilee** [comment]
You do not _need_ to use tolerance values if you dispose of the idea that you are looking at decimals. These are still binary numbers, despite their literal forms allowing decimals to be written or printed, which causes the following quirk: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b07964ddf88dbf040277d4292c509cec

But yes, I am otherwise joking. Except for the part where we can define what output we give if we implement `tan` ourselves: normally that function is offloaded to the system's math library.

**Comment by alexschrod** [comment]
I feel like one of the reasons `tan(pi / 2)` can't be `NaN` is because the true value of `pi / 2` can't be represented in floats, so you're never calculating that *exact* `tan` value. It's always going to be an approximation of it, which *doesn't* have an undefined/indeterminate value.

**Comment by yellowsink** [comment]
makes sense - if the after-float-error value of pi/2 lands below the real value in f64 and above the real value in f32 this makes logical sense

**Comment by workingjubilee** [comment]
Indeed, "my calculation has a bizarre result due to an intermediate step violating my expectations" is also why the standard recommends a profusion of tiny methods like atan2 and atan2pi and so on. The "pure trigonometry" answer that you may draw out by hand (and that a sophisticated program like Mathematica, which **can** do symbolic algebra, can arrive at) isn't the one that's fast to calculate, so if you are looking for a specific formula it can be worthwhile to go back to the drawing board and rearrange the subcomponents.

**Comment by quaternic** [comment]
> makes sense - if the after-float-error value of pi/2 lands below the real value in f64 and above the real value in f32 this makes logical sense

Yes, that's precisely the case here.

The value of std::f32::consts::PI is *exactly* `3.1415927410125732421875` (`= 13176795 * 2^-22`), which is approximately `0.0000000874...` (or  `0.367.. * 2^-22`) more than the real π. Using that rounded value, the correct tan would be `-22877332.42...` and `f32::tan` seems to return `-22877334.0`, which is within 1 ulp.

With some trig, you can derive `tan(π/2 + h) = -1/tan(h)`, so for some small `h`, `h = tan(h) = -1/tan(π/2 + h)`, which means that `-2.0 / (std::f32::consts::PI / 2.0).tan()` actually computes the *rounding error* of `std::f32::consts::PI`:
```
-2.0 / (std::f32::consts::PI / 2.0).tan() = 8.742278e-8
(std::f32::consts::PI as f64 - std::f64::consts::PI) as f32 = 8.742278e-8
```
(a simpler form of the exact same phenomenon is just sin(PI))

Anyway, I don't think there's any issue to fix here, other than *possibly* adding documentation that explicitly points out that the provided constants are rounded values because the real values are not representable. I don't think std needs this, but in some cases it might be useful to have a separate upper and lower bounds, so that if I want values in the range [0,π), I could consistently check `x <= PI_LB` instead of having to pick between `x < f32::PI` and `x <= f64::PI`

**Comment by Dante-Broggi** [comment]
As another language comparison, Swift [has `pi` round down for f32](https://github.com/apple/swift/blob/7a1829f601daa5a2b6d9ac252ba607a890848a13/stdlib/public/core/FloatingPointTypes.swift.gyb#L464-L470) to prevent this and similar issues. 
[Docs for `pi` in the `FloatingPoint ` protocol (trait)](https://github.com/apple/swift/blob/7a1829f601daa5a2b6d9ac252ba607a890848a13/stdlib/public/core/FloatingPoint.swift#L331-L342):
```swift
  /// The mathematical constant pi (π), approximately equal to 3.14159.
  /// 
  /// When measuring an angle in radians, π is equivalent to a half-turn.
  ///
  /// This value is rounded toward zero to keep user computations with angles
  /// from inadvertently ending up in the wrong quadrant. A type that conforms
  /// to the `FloatingPoint` protocol provides the value for `pi` at its best
  /// possible precision.
  ///
  ///     print(Double.pi)
  ///     // Prints "3.14159265358979"
  static var pi: Self { get }
```
